src/containers/downloads/sagas/downloadReferrals.js

```diff
@@ -0,0 +1,300 @@
+  dateReferred: 'Date Referred',
+  charge: 'Charge',
+  dateOfIntake: 'Date of Intake',
+  staff: 'CRC Staff Assigned',
+};
+
+function* downloadReferralsWorker(action :SequenceAction) :Saga<*> {
+
+  const { id } = action;
+  const workerResponse = {};
+
+  try {
+    yield put(downloadReferrals.request(id));
+    const { endDate, selectedAgency: agency, startDate } :Map = action.value;
+    const agencyName = getPropertyValue(agency, [NAME, 0]);
+    const agencyEKID :?UUID = getEntityKeyId(agency);
+    const startDateAsDateTime :DateTime = DateTime.fromISO(startDate);
+    const endDateAsDateTime :DateTime = DateTime.fromISO(endDate);
+
+    const agencyESID :UUID = yield select(selectEntitySetId(AGENCY));
+    const referralRequestESID :UUID = yield select(selectEntitySetId(REFERRAL_REQUEST));
+
+    let referralRequests :List = List();
+
+    if (agency.isEmpty()) {
+      if (!startDate.length && !endDate.length) {
+        const response = yield call(getEntitySetDataWorker, getEntitySetData({ entitySetId: referralRequestESID }));
+        if (response.error) throw response.error;
+        referralRequests = fromJS(response.data);
+      }
+      else {
+        const datetimeCompletedPTID :UUID = yield select(selectPropertyTypeId(DATETIME_COMPLETED));
+        const searchOptions = {
+          entitySetIds: [referralRequestESID],
+          start: 0,
+          maxHits: 10000,
+          constraints: [{
+            min: 1,
+            constraints: [{
+              searchTerm: getUTCDateRangeSearchString(
+                datetimeCompletedPTID,
+                'day',
+                startDateAsDateTime,
+                endDateAsDateTime
+              ),
+              fuzzy: false
+            }]
+          }]
+        };
+
+        const response = yield call(searchEntitySetDataWorker, searchEntitySetData(searchOptions));
+        if (response.error) throw response.error;
+        referralRequests = fromJS(response.data.hits);
+      }
+    }
+    else if (!agency.isEmpty()) {
+      const filter = {
+        entityKeyIds: [agencyEKID],
+        destinationEntitySetIds: [],
+        sourceEntitySetIds: [referralRequestESID],
+      };
+
+      const response :Object = yield call(
+        searchEntityNeighborsWithFilterWorker,
+        searchEntityNeighborsWithFilter({ entitySetId: agencyESID, filter })
+      );
+      if (response.error) throw response.error;
+      referralRequests = fromJS(response.data[agencyEKID]).map((neighbor :Map) => getNeighborDetails(neighbor));
+
+      if (startDate.length || endDate.length) {
+        referralRequests = referralRequests.filter((referralRequest :Map) => {
+          const referralDate = getPropertyValue(referralRequest, [DATETIME_COMPLETED, 0]);
+          const referralDateAsDateTime :DateTime = DateTime.fromISO(referralDate);
+          let referralWithinRange = true;
+          if (!startDateAsDateTime.isValid && !endDateAsDateTime.isValid) return referralWithinRange;
```

Comment:
I'm confused about this statement. If both start and end datetime values are invalid it seems the referralWithinRange value should be `false`... unless I'm missing something

Reply:
this is a very valid question. i decided to make the default behavior for the download to be to download all referrals (no matter the date), so here, if the datetime is invalid, then that means there was none selected, and so the behavior should be to just download all.